File: mid/6_Zigzag Conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows<=1:
            return s
        zigzag = ['' for i in range(numRows)]
        row, step = 0, 1
        
        for crct in s:
            zigzag[row] += crct

            if row == 0:
                step = 1
            elif row == numRows-1:
                step = -1
            
            row += step

        return ''.join(zigzag)

if __name__ == '__main__':
    s = "PAYPALISHIRING"
    numRows = 3
    ans = Solution().convert(s, numRows)
    print('ans=', ans)

File: mid/Array/6. Zigzag Conversion.cpp
class Solution {
public:
    string convert(string inputString, int numRows) {
        if (numRows <= 1) return inputString;
        
        vector<string> rows(min(numRows, int(inputString.size())), "");
        int direction = -1; // -1 represents down, 1 represents up
        int currentRow = 0;
        
        for (int i = 0; i < inputString.size(); i++) {
            rows[currentRow] += inputString[i];
            currentRow += (direction == -1) ? 1 : -1;
            
            if (currentRow == 0 || currentRow == numRows - 1) {
                direction = -direction;
            }
        }
        
        string result = "";
        for (const auto& row : rows) {
            result += row;
        }
        
        return result;
    }
};

int main(){
    string s = "PAYPALISHIRING";
    int numRows = 3;
    string ans = Solution().convert(s, numRows);
    cout << ans;

}